File: backend/src/modules/user/user.dto.ts
// src/modules/user/user.dto.ts
import { IsOptional, IsString, Length, IsNumber, IsBoolean } from 'class-validator';
import { ApiProperty } from '@nestjs/swagger';
import { Type } from 'class-transformer';;

// ==========================================
// [신규] 내 프로필 조회 응답용 DTO
// ==========================================

// 1. User 정보 객체 (export 권장)
export class UserProfileDto {
  @ApiProperty({ example: '550e8400-e29b-41d4-a716-446655440000' })
  id: string;

  @ApiProperty({ example: 'Lupin' })
  nickname: string;

  @ApiProperty({ example: 'lupin@gmail.com', description: '본인이므로 노출' })
  email: string;

  @ApiProperty({ example: 'https://s3.aws.com/profile/123.jpg', nullable: true })
  profileImage: string | null;

  @ApiProperty({ example: 'KAKAO' })
  provider: string;
}

// 2. Stat 정보 객체 (export 권장)
export class UserStatDto {
  @ApiProperty({ example: 1250 })
  policeMmr: number;

  @ApiProperty({ example: 1050 })
  thiefMmr: number;

  @ApiProperty({ example: 95 })
  integrityScore: number;

  @ApiProperty({ example: 15 })
  totalCatch: number;

  @ApiProperty({ example: 5 })
  totalRelease: number;

  @ApiProperty({ example: 3600, description: '초 단위' })
  totalSurvival: number;

  @ApiProperty({ example: 12.5, description: 'km 단위' })
  totalDistance: number;

  @ApiProperty({ example: 2 })
  totalMvpCount: number;
}

// 3. Achievement 정보 객체 (export 권장)
export class AchievementDto {
  @ApiProperty({ example: 'SPEED_DEMON_1' })
  achieveId: string;

  @ApiProperty({ example: '2026-01-20T10:00:00.000Z' })
  earnedAt: Date;
}

// ✅ [추가] Data 내부 구조를 정의하는 DTO 클래스
export class MyProfileDataDto {
  @ApiProperty({ type: UserProfileDto }) // $ref 대신 클래스 타입 사용
  user: UserProfileDto;

  @ApiProperty({ type: UserStatDto })
  stat: UserStatDto;

  @ApiProperty({ type: [AchievementDto] }) // 배열은 [Class] 형태로 지정
  achievements: AchievementDto[];
}

// 4. 최종 Response DTO (Wrapper)
export class MyProfileResponseDto {
  @ApiProperty({ example: true })
  success: boolean;

  @ApiProperty({ example: '내 프로필을 조회했습니다.' })
  message: string;

  // ✅ [수정] 복잡한 properties 설정 대신 type에 DTO 클래스 지정
  @ApiProperty({ type: MyProfileDataDto })
  data: MyProfileDataDto;

  @ApiProperty({ example: null, nullable: true })
  error: any;
}

// 1. 타인 User 정보 객체 (이메일 등 민감정보 제외)
export class OtherUserProfileDto {
  @ApiProperty({ example: 'target-uuid-5678' })
  id: string;

  @ApiProperty({ example: 'Sherlock' })
  nickname: string;

  @ApiProperty({ example: null, nullable: true })
  profileImage: string | null;

  @ApiProperty({ example: '2025-12-25T00:00:00Z', description: '가입일' })
  createdAt: Date;
}

// 2. Data 내부 구조 (Stat, Achievement는 기존 DTO 재사용 가능)
export class OtherUserProfileDataDto {
  @ApiProperty({ type: OtherUserProfileDto })
  user: OtherUserProfileDto;

  @ApiProperty({ type: UserStatDto }) // 기존에 만든 UserStatDto 재사용
  stat: UserStatDto;

  @ApiProperty({ type: [AchievementDto] }) // 기존 AchievementDto 재사용
  achievements: AchievementDto[];
}

// 3. 최종 Response DTO (Wrapper)
export class OtherProfileResponseDto {
  @ApiProperty({ example: true })
  success: boolean;

  @ApiProperty({ example: '유저 프로필을 조회했습니다.' })
  message: string;

  @ApiProperty({ type: OtherUserProfileDataDto })
  data: OtherUserProfileDataDto;

  @ApiProperty({ example: null, nullable: true })
  error: any;
}
// ✅ [수정] Request DTO: example 추가
export class UpdateProfileDto {
  @ApiProperty({ 
    example: 'NewLupin', // "string" 대신 실제 예시 값 표시
    description: '변경할 닉네임', 
    required: false 
  })
  @IsOptional()
  @IsString()
  @Length(2, 12)
  nickname?: string;

  @ApiProperty({ 
    example: 'https://new-image-url.com/face.png', // 실제 예시 값 표시
    description: '변경할 프로필 이미지 URL', 
    required: false 
  })
  @IsOptional()
  @IsString()
  profileImage?: string;
}

// ==========================================
// [신규] 프로필 수정 응답용 DTO
// ==========================================

class UpdateProfileDataDto {
  @ApiProperty({ example: 'NewLupin' })
  nickname: string;

  @ApiProperty({ example: '2026-01-24T16:00:00Z' })
  updatedAt: Date;
}

export class UpdateProfileResponseDto {
  @ApiProperty({ example: true })
  success: boolean;

  @ApiProperty({ example: '프로필이 수정되었습니다.' })
  message: string;

  @ApiProperty({ type: UpdateProfileDataDto })
  data: UpdateProfileDataDto;

  @ApiProperty({ example: null, nullable: true })
  error: any;
}

// 1. 전적 조회 쿼리 DTO
export class MatchHistoryQueryDto {
  @ApiProperty({ description: '페이지 번호 (기본값: 1)', required: false })
  @IsOptional()
  @Type(() => Number)
  @IsNumber()
  page?: number = 1;

  @ApiProperty({ description: '한 번에 가져올 개수 (기본값: 10)', required: false })
  @IsOptional()
  @Type(() => Number)
  @IsNumber()
  limit?: number = 10;
}

// 1. 통계 정보
class MyStatDto {
  @ApiProperty({ example: 3 })
  catchCount: number;

  @ApiProperty({ example: 85 })
  contribution: number;
}

// 2. 맵 설정 정보 (JSON 구조 반영)
class MapConfigDto {
  @ApiProperty({ example: [{ lat: 37.123, lng: 127.123 }] })
  polygon: object[];

  @ApiProperty({ example: { lat: 37.1235, lng: 127.1235, radiusM: 12 } })
  jail: object;
}

// 3. 게임 규칙 정보 (복잡한 JSON 예시 반영)
class GameRulesDto {
  @ApiProperty({ example: 'NON_CONTACT' })
  contactMode: string;

  @ApiProperty({ 
    example: { 
      ruleType: 'THREE_OF_THREE', 
      nearThresholdM: 1.0, 
      speedMaxMps: 1.2 
    } 
  })
  captureRule: object;

  @ApiProperty({ 
    example: { 
      jailEnabled: true, 
      rescue: { type: 'CHANNELING', rangeM: 10 } 
    } 
  })
  jailRule: object;
}

// 4. 게임 상세 정보
class GameInfoDto {
  @ApiProperty({ example: 8 })
  maxPlayers: number;

  @ApiProperty({ example: 600, description: '제한 시간 (초)' })
  timeLimit: number;

  @ApiProperty({ type: MapConfigDto })
  mapConfig: MapConfigDto;

  @ApiProperty({ example: 540, description: '실제 진행 시간 (초)' })
  playTime: number;

  @ApiProperty({ example: '2026-01-25T14:00:00Z' })
  playedAt: Date;

  @ApiProperty({ type: GameRulesDto })
  rules: GameRulesDto;
}

// 5. 전적 아이템 (배열 내부 요소)
class MatchRecordDto {
  @ApiProperty({ example: 'uuid-1234' })
  matchId: string;

  @ApiProperty({ example: 'WIN' })
  result: string;

  @ApiProperty({ example: 'POLICE' })
  role: string;

  @ApiProperty({ type: MyStatDto })
  myStat: MyStatDto;

  @ApiProperty({ type: GameInfoDto })
  gameInfo: GameInfoDto;
}

// 6. 최종 Response DTO
export class MatchHistoryResponseDto {
  @ApiProperty({ example: true })
  success: boolean;

  @ApiProperty({ example: '전적 기록을 조회했습니다.' })
  message: string;

  @ApiProperty({ type: [MatchRecordDto] }) // 배열 형태 지정
  data: MatchRecordDto[];

  @ApiProperty({ example: null, nullable: true })
  error: any;
}

// ==========================================
// [수정] 회원 탈퇴 요청 DTO (Request)
// ==========================================
export class DeleteAccountDto {
  @ApiProperty({ 
    example: '게임이 너무 어려워요', // ✅ "string" 대신 예시 값 표시
    description: '탈퇴 사유 (선택)', 
    required: false 
  })
  @IsOptional()
  @IsString()
  reason?: string;

  @ApiProperty({ 
    example: true, // ✅ 예시 값 표시
    description: '데이터 삭제 동의 여부 (true여야 탈퇴 가능)', 
    required: true 
  })
  @IsBoolean()
  agreedToLoseData: boolean;
}

// ==========================================
// [신규] 회원 탈퇴 응답용 DTO (Response)
// ==========================================
export class DeleteAccountResponseDto {
  @ApiProperty({ example: true })
  success: boolean;

  @ApiProperty({ example: '회원 탈퇴가 완료되었습니다. 이용해 주셔서 감사합니다.' })
  message: string;

  @ApiProperty({ example: null, nullable: true })
  data: any; // 데이터는 null

  @ApiProperty({ example: null, nullable: true })
  error: any;
}

File: frontend/lib/data/dto/user_dto.dart
import 'package:json_annotation/json_annotation.dart';
import 'auth_dto.dart'; // UserDto

part 'user_dto.g.dart';

// ------------------------------------------------------------------
// My Profile
// ------------------------------------------------------------------

@JsonSerializable()
class UserStatDto {
  final int? policeMmr;
  final int? thiefMmr;
  final int totalCatch;
  final int totalSurvival;
  final double totalDistance;
  final int? integrityScore;
  final int? totalRelease;
  final int? totalMvpCount;

  UserStatDto({
    this.policeMmr,
    this.thiefMmr,
    required this.totalCatch,
    required this.totalSurvival,
    required this.totalDistance,
    this.integrityScore,
    this.totalRelease,
    this.totalMvpCount,
  });

  factory UserStatDto.fromJson(Map<String, dynamic> json) =>
      _$UserStatDtoFromJson(json);
  Map<String, dynamic> toJson() => _$UserStatDtoToJson(this);
}

@JsonSerializable()
class AchievementDto {
  final String achieveId;
  final DateTime earnedAt;

  AchievementDto({required this.achieveId, required this.earnedAt});

  factory AchievementDto.fromJson(Map<String, dynamic> json) =>
      _$AchievementDtoFromJson(json);
  Map<String, dynamic> toJson() => _$AchievementDtoToJson(this);
}

@JsonSerializable()
class MyProfileDataDto {
  final UserDto user;
  final UserStatDto stat;
  final List<AchievementDto> achievements;

  MyProfileDataDto({
    required this.user,
    required this.stat,
    required this.achievements,
  });

  factory MyProfileDataDto.fromJson(Map<String, dynamic> json) =>
      _$MyProfileDataDtoFromJson(json);
  Map<String, dynamic> toJson() => _$MyProfileDataDtoToJson(this);
}

@JsonSerializable()
class MyProfileResponseDto {
  final bool? success;
  final String? message;
  final MyProfileDataDto? data;
  final dynamic error;

  MyProfileResponseDto({this.success, this.message, this.data, this.error});

  factory MyProfileResponseDto.fromJson(Map<String, dynamic> json) =>
      _$MyProfileResponseDtoFromJson(json);
}

// ------------------------------------------------------------------
// Other Profile
// ------------------------------------------------------------------

@JsonSerializable()
class OtherUserProfileDto {
  final String id;
  final String nickname;
  final String? profileImage;
  final DateTime createdAt;

  OtherUserProfileDto({
    required this.id,
    required this.nickname,
    this.profileImage,
    required this.createdAt,
  });

  factory OtherUserProfileDto.fromJson(Map<String, dynamic> json) =>
      _$OtherUserProfileDtoFromJson(json);
  Map<String, dynamic> toJson() => _$OtherUserProfileDtoToJson(this);
}

@JsonSerializable()
class OtherUserProfileDataDto {
  final OtherUserProfileDto user;
  final UserStatDto stat;
  final List<AchievementDto> achievements;

  OtherUserProfileDataDto({
    required this.user,
    required this.stat,
    required this.achievements,
  });

  factory OtherUserProfileDataDto.fromJson(Map<String, dynamic> json) =>
      _$OtherUserProfileDataDtoFromJson(json);
}

@JsonSerializable()
class OtherProfileResponseDto {
  final bool? success;
  final String? message;
  final OtherUserProfileDataDto? data;
  final dynamic error;

  OtherProfileResponseDto({this.success, this.message, this.data, this.error});

  factory OtherProfileResponseDto.fromJson(Map<String, dynamic> json) =>
      _$OtherProfileResponseDtoFromJson(json);
}

// ------------------------------------------------------------------
// Update Profile
// ------------------------------------------------------------------

@JsonSerializable()
class UpdateProfileDto {
  final String? nickname;
  final String? profileImage;

  UpdateProfileDto({this.nickname, this.profileImage});

  Map<String, dynamic> toJson() => _$UpdateProfileDtoToJson(this);
}

@JsonSerializable()
class UpdateProfileDataDto {
  final String nickname;
  final DateTime updatedAt;

  UpdateProfileDataDto({required this.nickname, required this.updatedAt});

  factory UpdateProfileDataDto.fromJson(Map<String, dynamic> json) =>
      _$UpdateProfileDataDtoFromJson(json);
}

@JsonSerializable()
class UpdateProfileResponseDto {
  final bool? success;
  final String? message;
  final UpdateProfileDataDto? data;
  final dynamic error;

  UpdateProfileResponseDto({this.success, this.message, this.data, this.error});

  factory UpdateProfileResponseDto.fromJson(Map<String, dynamic> json) =>
      _$UpdateProfileResponseDtoFromJson(json);
}

// ------------------------------------------------------------------
// Match History
// ------------------------------------------------------------------

@JsonSerializable()
class MatchHistoryQueryDto {
  final int page;
  final int limit;

  MatchHistoryQueryDto({this.page = 1, this.limit = 10});

  Map<String, dynamic> toJson() => _$MatchHistoryQueryDtoToJson(this);
}

@JsonSerializable()
class MyStatDto {
  final int catchCount;
  final int? rescueCount; // Added for Thief stats
  final int contribution;

  MyStatDto({
    required this.catchCount,
    this.rescueCount,
    required this.contribution,
  });

  factory MyStatDto.fromJson(Map<String, dynamic> json) =>
      _$MyStatDtoFromJson(json);
  Map<String, dynamic> toJson() => _$MyStatDtoToJson(this);
}

@JsonSerializable()
class MapConfigDto {
  final List<dynamic> polygon; // Or specific PointDto
  final dynamic jail; // Or JailDto

  MapConfigDto({required this.polygon, required this.jail});

  factory MapConfigDto.fromJson(Map<String, dynamic> json) =>
      _$MapConfigDtoFromJson(json);
  Map<String, dynamic> toJson() => _$MapConfigDtoToJson(this);
}

@JsonSerializable()
class GameRulesDto {
  final String contactMode;
  final dynamic captureRule;
  final dynamic jailRule;

  GameRulesDto({
    required this.contactMode,
    required this.captureRule,
    required this.jailRule,
  });

  factory GameRulesDto.fromJson(Map<String, dynamic> json) =>
      _$GameRulesDtoFromJson(json);
  Map<String, dynamic> toJson() => _$GameRulesDtoToJson(this);
}

@JsonSerializable()
class GameInfoDto {
  final int maxPlayers;
  final int timeLimit;
  final MapConfigDto mapConfig;
  final int playTime;
  final DateTime playedAt;
  final GameRulesDto rules;

  GameInfoDto({
    required this.maxPlayers,
    required this.timeLimit,
    required this.mapConfig,
    required this.playTime,
    required this.playedAt,
    required this.rules,
  });

  factory GameInfoDto.fromJson(Map<String, dynamic> json) =>
      _$GameInfoDtoFromJson(json);
  Map<String, dynamic> toJson() => _$GameInfoDtoToJson(this);
}

@JsonSerializable()
class MatchRecordDto {
  final String matchId;
  final String result; // WIN, LOSE, DRAW
  final String role; // POLICE, THIEF
  final MyStatDto myStat;
  final GameInfoDto gameInfo;

  MatchRecordDto({
    required this.matchId,
    required this.result,
    required this.role,
    required this.myStat,
    required this.gameInfo,
  });

  factory MatchRecordDto.fromJson(Map<String, dynamic> json) =>
      _$MatchRecordDtoFromJson(json);
  Map<String, dynamic> toJson() => _$MatchRecordDtoToJson(this);
}

@JsonSerializable()
class MatchHistoryResponseDto {
  final bool? success;
  final String? message;
  final List<MatchRecordDto>? data;
  final dynamic error;

  MatchHistoryResponseDto({this.success, this.message, this.data, this.error});

  factory MatchHistoryResponseDto.fromJson(Map<String, dynamic> json) =>
      _$MatchHistoryResponseDtoFromJson(json);
}

// ------------------------------------------------------------------
// Delete Account
// ------------------------------------------------------------------

@JsonSerializable()
class DeleteAccountDto {
  final String? reason;
  final bool agreedToLoseData;

  DeleteAccountDto({this.reason, required this.agreedToLoseData});

  Map<String, dynamic> toJson() => _$DeleteAccountDtoToJson(this);
}

@JsonSerializable()
class DeleteAccountResponseDto {
  final bool? success;
  final String? message;
  final dynamic data;
  final dynamic error;

  DeleteAccountResponseDto({this.success, this.message, this.data, this.error});

  factory DeleteAccountResponseDto.fromJson(Map<String, dynamic> json) =>
      _$DeleteAccountResponseDtoFromJson(json);
}
